common/migration_mechanism.py

```diff
@@ -0,0 +1,192 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2018  Joseph Spearritt <mcoot@tamods.org>
+#
+# This file is part of taserver
+#
+# taserver is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# taserver is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with taserver.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+from typing import Dict, List
+
+from collections import OrderedDict
+from functools import wraps
+import os
+import shutil
+import json
+import copy
+import glob
+
+
+# Known migrations
+_registered_migrations = OrderedDict()
+
+
+def _perform_backups(data_root: str, schema_version: int) -> None:
+    backup_path = data_root + ('.%d.bkp' % schema_version)
+    if os.path.exists(backup_path):
+        shutil.rmtree(backup_path)
```

Comment:
I think we're going to lose our backups if migrations ever fail and we try it a second time.

Reply:
Fixed. It now no longer puts the schema version (pointless given it will backup metadata.json), and numbers backups using the first available number, never overwriting.